lib/ansible/modules/windows/win_hostname.ps1

```diff
@@ -0,0 +1,54 @@
+#!powershell
+
+# WANT_JSON
+# POWERSHELL_COMMON
+
+########
+
+function Get-ComputerName() {
+    Get-Content env:computername
+    return
+}
+
+function Set-ComputerName($newComputerName) {
+    $systemInfo = Get-WmiObject Win32_ComputerSystem
+    $cmdResult = $systemInfo.Rename($newComputerName)    
+    if ($cmdResult.ReturnValue -ne 0) {         
+        Fail-Json $result "The computer name could not be set. Check that it is a valid name."                
+    }
+    return $cmdResult.ReturnValue
+}
+
+########
+
+$params = Parse-Args $args -supports_check_mode $true;
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+$diff_mode = Get-AnsibleParam -obj $params -name "_ansible_diff" -type "bool" -default $false
+
+$result = New-Object psobject @{
+    changed = $false
+};
+
+If (-not $params.name.GetType)
+{
+    Fail-Json $result "missing required arguments: name"
+}
+
+$newComputerName = Get-Attr $params "name"
+$currentComputerName = (Get-ComputerName)
+
+if ($check_mode) {
+  $result.msg = "Running in check mode - would change computer name to $newComputerName from $currentComputerName"  
+}
+if (-not $check_mode) {
+    if ($newComputerName -ne $currentComputerName) {
+        $setResult = (Set-ComputerName($newComputerName))
```

Comment:
I am not a Windows/Powershell specialist, but why are we not using `Rename-Computer` here ?
This cmdlet supports `-WhatIf`

Reply:
Rename-Computer is available from Powershell 3.0

Reply:
PowerShell 3.0 and above is used by most, if not all Windows modules anyway.

Reply:
Yes you should be using `Rename-Computer` here, it has some extra features like `-WhatIf` support and does the validation of the result for you, return checking when it calls the WMI object internally for you. While yes it was only added from 3.0, that is a minimum requirement for Ansible and there is no reason why we should avoid using cmdlets added in that version.

In this case you can easily simplify this whole module with

```
# Requires -Module Ansible.ModuleUtils.Legacy

$params = Parse-Args $args -supports_check_mode $true
$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false

$name = Get-AnsibleParam -obj $params -name "name" -type "str" -failifempty $true

$result = @{
    changed = $false
}

$current_name = $env:computername
if ($name -ne $current_name) {
    Rename-Computer -NewName $name -WhatIf:$check_mode -Force:$true -Restart:$false
    $result.changed = $true
}

Exit-Json -obj $result
```

No messing around with WMI, checking return values, no if statements that control the check mode branches and the like and so on.